scipy/special/_logsumexp.py

```diff
@@ -100,53 +103,117 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     else:
-        real = xp.real(a) if xp.isdtype(a.dtype, "complex floating") else a
-        a_max = xp.max(real, axis=axis, keepdims=True)
+        max = xp.max(a, axis=axis, keepdims=True)
+        mask = a == max
 
-    if a_max.ndim > 0:
-        a_max[~xp.isfinite(a_max)] = 0
-    elif not xp.isfinite(a_max):
-        a_max = 0
+    return xp.asarray(max), xp.asarray(mask)
 
-    if b is not None:
-        b = xp.asarray(b)
-        tmp = b * xp.exp(a - a_max)
-    else:
-        tmp = xp.exp(a - a_max)
 
-    # suppress warnings about log of zero
-    with np.errstate(divide='ignore'):
-        s = xp.sum(tmp, axis=axis, keepdims=keepdims, dtype=tmp.dtype)
-        if return_sign:
-            # For complex, use the numpy>=2.0 convention for sign.
-            if xp.isdtype(s.dtype, "complex floating"):
-                sgn = s / xp.where(s == 0, xp.asarray(1, dtype=s.dtype), xp.abs(s))
-            else:
-                sgn = xp.sign(s)
-            s = xp.abs(s)
-        out = xp.log(s)
-
-    if not keepdims:
-        a_max = xp.squeeze(a_max, axis=axis)
-    out += a_max
+def _sign(x, xp):
+    return x / xp.where(x == 0, xp.asarray(1, dtype=x.dtype), xp.abs(x))
+
 
+def _logsumexp(a, b, axis, return_sign, xp):
+
+    # This has been around for about a decade, so let's consider it a feature:
+    # Even if element of `a` is infinite or NaN, it adds nothing to the sum if
+    # the corresponding weight is zero.
+    if b is not None:
+        a[b == 0] = -xp.inf
+
+    # Find element with maximum real part, since this is what affects the magnitude
+    # of the exponential. Possible enhancement: include log of `b` magnitude in `a`.
+    a_max, i_max = _elements_and_indices_with_max_real(a, axis=axis, xp=xp)
+
+    # for precision, these terms are separated out of the main sum.
+    a[i_max] = -xp.inf
+    i_max_dt = xp.astype(i_max, a.dtype)
+    # This is an inefficient way of getting `m` because it is the sum of a sparse
+    # array; however, this is the simplest way I can think of to get the right shape.
+    m = (xp.sum(i_max_dt, axis=axis, keepdims=True, dtype=a.dtype) if b is None
+         else xp.sum(b * i_max_dt, axis=axis, keepdims=True, dtype=a.dtype))
+
+    # Arithmetic between infinities will introduce NaNs.
+    # `+ a_max` at the end naturally corrects for removing them here.
+    shift = xp.where(xp.isfinite(a_max), a_max, xp.asarray(0, dtype=a_max.dtype))
+
+    # Shift, exponentiate, scale, and sum
+    exp = b * xp.exp(a - shift) if b is not None else xp.exp(a - shift)
+    s = xp.sum(exp, axis=axis, keepdims=True, dtype=exp.dtype)
+    s = xp.where(s == 0, s, s/m)
+
+    # Separate sign/magnitude information
+    sgn = None
     if return_sign:
-        return out, sgn
-    else:
-        return out
+        # Use the numpy>=2.0 convention for sign.
+        # When all array libraries agree, this can become sng = xp.sign(s).
+        sgn = _sign(s + 1, xp=xp) * _sign(m, xp=xp)
```

Comment:
Is there an array-api-compat issue for this?

Reply:
I didn't actually check whether `xp.sign` can just be used directly because the code was already branching for this. Let me check. And if it can't and there is no issue, I'll open one.

Reply:
Am I doing something wrong? It looks like only NumPy 2.0 and JAX work properly here, and the definition for complex numbers has been in the standard since 2022.12.

```python3
from array_api_compat import numpy as np
np.sign((2 + 5j))  # (1+0j) (NumPy 1.26.4)

from array_api_compat import torch as xp
xp.sign(xp.asarray((2 + 5j)))
# RuntimeError: Unlike NumPy, torch.sign is not intended to support complex numbers. Please use torch.sgn instead.

import array_api_strict as xp
xp.sign(xp.asarray((2 + 5j)))  # Array(1.+0.j, dtype=array_api_strict.complex128)
```

There is no bug report at array-api-compat or array-api-strict. Do you agree I should make one?

Reply:
https://github.com/data-apis/array-api-compat/issues/136, so we should be able to use `xp.sign` after updating array-api-compat. Happy to keep the shim for this PR if you don't want to update the submodule here, but if so could you move it to `_lib` as `xp_sign` so that we remember to remove it?

Maybe the bug is still present in array-api-strict.

Reply:
The PR that closed that issue didn't fix NumPy pre 2.0, though.

Reply:
> but if so could you move it

We already have an `xp_sign` there. I don't really want to combine the two.

Reply:
Okay cool, let's merge this and you can open those upstream issues.